pyo3-ffi/README.md

```diff
@@ -0,0 +1,188 @@
+This crate provides [Rust](https://www.rust-lang.org/) FFI declarations for Python 3.
+It supports both the stable and the unstable component of the ABI trough the use of cfg flags.
+Python Versions 3.7+ are supported.
+It is meant for advanced users only - regular PyO3 users shouldn't
+need to interact with this crate at all.
+
+The contents of this crate are not documented here, as it would entail
+basically copying the documentation from CPython. Consult the [Python/C API Reference
+Manual][capi] for up-to-date documentation.
+
+# Minimum supported Rust and Python versions
+
+PyO3 supports the following software versions:
+  - Python 3.7 and up (CPython and PyPy)
+  - Rust 1.48 and up
+
+# Example: Building Python Native modules
+
+PyO3 can be used to generate a native Python module. The easiest way to try this out for the
+first time is to use [`maturin`]. `maturin` is a tool for building and publishing Rust-based
+Python packages with minimal configuration. The following steps set up some files for an example
+Python module, install `maturin`, and then show how to build and import the Python module.
+
+First, create a new folder (let's call it `string_sum`) containing the following two files:
+
+**`Cargo.toml`**
+
+```toml
+[lib]
+name = "string_sum"
+# "cdylib" is necessary to produce a shared library for Python to import from.
+#
+# Downstream Rust code (including code in `bin/`, `examples/`, and `tests/`) will not be able
+# to `use string_sum;` unless the "rlib" or "lib" crate type is also included, e.g.:
+# crate-type = ["cdylib", "rlib"]
+crate-type = ["cdylib"]
+
+[dependencies.pyo3-ffi]
+orkaround for `extended_key_value_attributes`: https://github.com/rust-lang/rust/issues/82768#issuecomment-803935643
+fg_attr(docsrs, cfg_attr(docsrs, doc = concat!("version = \"", env!("cargo_pkg_version"),  "\"")))]
+fg_attr(not(docsrs), doc = "version = \"*\"")]
+features = ["extension-module"]
+```
+
+**`src/lib.rs`**
+```rust
+use std::intrinsics::transmute;
+use std::os::raw::c_char;
+
+use pyo3_ffi::*;
+
+#[allow(non_snake_case)]
+#[no_mangle]
+pub unsafe extern "C" fn PyInit_string_sum() -> *mut PyObject {
+    let init = PyModuleDef {
+        m_base: PyModuleDef_HEAD_INIT,
+        m_name: "string_sum\0".as_ptr() as *const c_char,
+        m_doc: std::ptr::null(),
+        m_size: 0,
+        m_methods: std::ptr::null_mut(),
+        m_slots: std::ptr::null_mut(),
+        m_traverse: None,
+        m_clear: None,
+        m_free: None,
+    };
+
+    let mptr = PyModule_Create(Box::into_raw(Box::new(init)));
+    let version = env!("CARGO_PKG_VERSION");
+    PyModule_AddObject(
+        mptr,
+        "__version__\0".as_ptr() as *const c_char,
+        PyUnicode_FromStringAndSize(version.as_ptr() as *const c_char, version.len() as isize),
+    );
+
+    let wrapped_sum_as_string = PyMethodDef {
+        ml_name: "sum_as_string\0".as_ptr() as *const c_char,
+        ml_meth: Some(transmute::<_PyCFunctionFast, PyCFunction>(sum_as_string)),
```

Comment:
This transmute looks a bit fishy, can you add a comment explaining why this is sound? Transmuting function pointers (and pyo3 users copypasting unsafe code they might not understand) always makes me nervous. This also fails to compile.

Reply:
Sure I will add a comment. Essentially this is save because we add the appropriate [`METH_FASTCALL`](https://docs.python.org/3/c-api/structures.html#METH_FASTCALL) flag below. Essentially there are different types of functions signatures accepted by python (which does the exact same pointer cast just backwards when the flag is specified).
In regards to your compiler failure: I am guessing that you are either using master pyo3 or didn't pull my newest changes. I changed the defintion of `_PyCFunctionFast` in the FFI bindings because they were actually incorrect (see above and compare the changes I made to the defintion to the one in the [python docs](https://docs.python.org/3/c-api/structures.html#c._PyCFunctionFast)). If this is not the cause could you send me your exact error because I actually tested and ran this example locally (and it worked :D)

Reply:
> I changed the defintion of _PyCFunctionFast in the FFI bindings

Ah, I was just importing `pyo3::fii::*;` from main, that makes sense.